Make labels, fields expose selectable requirements

hongchaodeng opened the pull request:
What?
This is to change the labels/fields Selector interface and make them expose selectable requirements. We reuse labels.Requirement struct for label selector and add fields.Requirement for field selector.

Why?
In order to index labels/fields, we need them to tell us three things: index key (a field or a label), operator (greater, less, or equal), and value (string, int, etc.). By getting selectable requirements, we are able to pass them down and use them for indexing in storage layer.

<!-- Reviewable:start -->

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/30658)

<!-- Reviewable:end -->

hongchaodeng committed:
make labels, fields expose selectable requirements

xiang90 commented on pkg/fields/selector.go:
```diff
@@ -39,6 +42,10 @@ type Selector interface {
 	// applied to the entire selector, or an error if fn returns an error.
 	Transform(fn TransformFunc) (Selector, error)
 
+	// ToRequirements converts this interface into Requirement struct to expose
```
// Requirements returns the label requirements of the selector.

xiang90 replied:
ToRequirements -> Requirements.

xiang90 commented on pkg/fields/selector.go:
```diff
@@ -75,6 +82,14 @@ func (t *hasTerm) Transform(fn TransformFunc) (Selector, error) {
+func (t *hasTerm) ToRequirements() []labels.Requirement {
+	r, err := labels.NewRequirement(t.field, labels.EqualsOperator, sets.NewString(t.value))
+	if err != nil {
+		panic(fmt.Sprintf("Impossible to create invalid requirement label. Err: %v", err))
```
does glog has a panic func?

hongchaodeng replied:
glog.Fatal() does similar job. Shall we import glog?

xiang90 replied:
probably not...

xiang90 commented on pkg/fields/selector.go:
```diff
@@ -157,6 +180,14 @@ func (t andTerm) Transform(fn TransformFunc) (Selector, error) {
 	return andTerm(next), nil
 }
 
+func (t andTerm) ToRequirements() (reqs []labels.Requirement) {
```
remove the named return. preallocate the slice?

xiang90 commented on pkg/labels/selector.go:
```diff
@@ -41,6 +41,12 @@ type Selector interface {
 	// Add adds requirements to the Selector
 	Add(r ...Requirement) Selector
+
+	// Requirements converts this interface into Requirement struct to expose
```
// Requirements returns the label requirements of the selector.

hongchaodeng commented:
cc @lavalamp

hongchaodeng commented:
ref: #29888

lavalamp commented on pkg/fields/selector.go:
```diff
@@ -75,6 +82,14 @@ func (t *hasTerm) Transform(fn TransformFunc) (Selector, error) {
+func (t *hasTerm) Requirements() []labels.Requirement {
+	r, err := labels.NewRequirement(t.field, labels.EqualsOperator, sets.NewString(t.value))
+	if err != nil {
+		panic(fmt.Sprintf("Impossible to create invalid requirement label. Err: %v", err))
```
if it's impossible, then why does that method return an error?

hongchaodeng replied:
NewRequirement validates input. In this case, EqualsOperator only allows `len(values) = 1`, which we explicitly ensure.

I think this is not valid either. Field syntax is different from label syntax. We should have another method to convert field selector into Requirements. Will change it shortly.

hongchaodeng commented:
@lavalamp @smarterclayton 
Updated the PR and top description.

I added fields.Requirement instead of reusing labels.Requirement because labels and fields have very different meaning and have been segregated from reading the code. Instead of refactoring and reusing constructor entry, I would say having a separate fields.Requirement is simpler and cleaner.

hongchaodeng commented:
@lavalamp PTAL and I would like to hear more feedback from you. Thanks!

lavalamp commented on pkg/fields/requirements.go:
```diff
@@ -0,0 +1,27 @@
+// Requirement contains a field, a value, and an operator that relates the field and value.
+// This is currently for reading internal selection information of field selector.
+type Requirement struct {
+	Operator labels.Operator
```
don't import the label package?

And/or implement the Requirement concept in a 3rd package.

lavalamp commented on pkg/labels/selector.go:
```diff
@@ -301,6 +307,8 @@ func (lsel internalSelector) Matches(l Labels) bool {
 	return true
 }
 
+func (lsel internalSelector) Requirements() ([]Requirement, bool) { return lsel, true }
```
Somewhere, name the []Requirement type and document that it's an implicit AND of all the elements?

hongchaodeng commented:
@lavalamp 
I move Operator into a new package "pkg/selection/" for reuse. I think converging Requirements of labels and fields is not a good option here, since they are segregated. It would be great to keep fields in private for label Requirement.

I add `Requirements` type and document it "AND of all reqs".

PTAL. Thanks!

hongchaodeng commented:
@lavalamp 
Let me know your thoughts and I can iterate again

lavalamp commented on pkg/selection/operator.go:
```diff
@@ -0,0 +1,33 @@
+type Operator string
+
+const (
+	DoesNotExistOperator Operator = "!"
```
since this is a tight, focused package, consider omitting the "Operator" suffix on all of these. Will make usage much less verbose..

lavalamp commented:
just one nit from me.

hongchaodeng commented:
@lavalamp 
Addressed comments. Tests passed.
Can you take another look?

lavalamp commented:
lgtm

hongchaodeng commented:
Rebased to fix one new conflict:
```
pkg/registry/generic/registry/store_test.go:91: undefined: "k8s.io/kubernetes/pkg/labels".InOperator
```
